src/main/java/org/cloudburstmc/server/command/defaults/SetBlockCommand.java

```diff
@@ -0,0 +1,114 @@
+package org.cloudburstmc.server.command.defaults;
+
+import com.nukkitx.math.vector.Vector3i;
+import com.nukkitx.protocol.bedrock.data.command.CommandParamType;
+import org.cloudburstmc.server.block.BlockState;
+import org.cloudburstmc.server.block.BlockStates;
+import org.cloudburstmc.server.block.util.BlockStateMetaMappings;
+import org.cloudburstmc.server.command.Command;
+import org.cloudburstmc.server.command.CommandSender;
+import org.cloudburstmc.server.command.CommandUtils;
+import org.cloudburstmc.server.command.data.CommandData;
+import org.cloudburstmc.server.command.data.CommandParameter;
+import org.cloudburstmc.server.locale.TranslationContainer;
+import org.cloudburstmc.server.player.Player;
+import org.cloudburstmc.server.registry.BlockRegistry;
+import org.cloudburstmc.server.utils.Identifier;
+import org.cloudburstmc.server.utils.TextFormat;
+
+public class SetBlockCommand extends Command {
+
+    public SetBlockCommand() {
+        super("setblock", CommandData.builder("setblock")
+                .setDescription("commands.setblock.description")
+                .setUsageMessage("/setblock <position> <tileName> [tileData] [replace|destroy|keep]")
+                .setPermissions("cloudburst.command.setblock")
+                .setParameters(
+                        new CommandParameter[]{
+                                new CommandParameter("position", CommandParamType.BLOCK_POSITION, false),
+                                new CommandParameter("Block", false, BlockRegistry.get().getBlockStates().stream().map(b -> b.getType().getName()).distinct().toArray(String[]::new)),
```

Comment:
There exists a client side enum for this already, no need to waste bandwidth sending every block name to every client.

The command parameter api will need to be changed slightly to allow custom enum names (although this could cause plugin clashes, but still it’s needed to use an enum name Item or Block to get the client side suggestions (with icons!))

Reply:
hm but vanilla also sends all these values

Reply:
It does but vanilla also sets the enum name to Block (that’s how the block icons show in the autocompletion).

But remember does a lot of stupid stuff as well. Might be useful to test if sending the enum values actual changes anything when the enum name is Block or Item

Reply:
I agree with @lukeeey. We'd also be sending significantly more data in the AvailableCommandsPacket with this too.